tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -877,56 +879,111 @@ static void setLocationInfoForClangNode(ClangNode ClangNode,
-    if (auto GeneratedSourceInfo = SM.getGeneratedSourceInfo(DeclBufID)) {
-      if (GeneratedSourceInfo->kind ==
-          GeneratedSourceInfo::ReplacedFunctionBody) {
-        // The location was in a temporary source buffer that just contains the
-        // function body and which we created while reusing the ASTContext for
-        // the rest of the file. Map the location back to the original file.
-        unsigned OriginalBufID = SM.findBufferContainingLoc(
-            GeneratedSourceInfo->originalSourceRange.getStart());
-        auto OriginalStartOffset = SM.getLocOffsetInBuffer(
-            GeneratedSourceInfo->originalSourceRange.getStart(), OriginalBufID);
-        auto GeneratedStartOffset = SM.getLocOffsetInBuffer(
-            GeneratedSourceInfo->generatedSourceRange.getStart(), DeclBufID);
-        Location.Offset += OriginalStartOffset - GeneratedStartOffset;
-        std::tie(Location.Line, Location.Column) =
-            SM.getPresumedLineAndColumnForLoc(DeclLoc, DeclBufID);
+      if (auto FDR = FD->getParameterListSourceRange())
+        range = FDR;
+    }
+
+    unsigned bufID = SM.findBufferContainingLoc(range.Start);
+
+    // If this range is from a generated buffer, recursively "unexpand" macros
+    // to chase after where the macro was originally expanded from.
+    //
+    // However, we don't care about certain kinds of generated buffers, so save
+    // the original range and buffer ID so we can set location according to VD.
+    auto VDRange = range;
+    auto VDBufID = bufID;
+    while (auto *info = SM.getGeneratedSourceInfo(bufID)) {
+      switch (info->kind) {
+#define MACRO_ROLE(Name, Description)                                          \
+  case GeneratedSourceInfo::Name##MacroExpansion:
+#include "swift/Basic/MacroRoles.def"
+        if (auto *customAttr = info->attachedMacroCustomAttr)
+          range = customAttr->getRange();
+        else
+          range = ASTNode::getFromOpaqueValue(info->astNode).getSourceRange();
+        bufID = SM.findBufferContainingLoc(range.Start);
+        continue; // Continue while-loop to recursively un-expand macros
+
+      case GeneratedSourceInfo::ReplacedFunctionBody:
+        if (bufID == VDBufID) {
+          // The location was in a temporary source buffer that just contains
+          // the function body, which we created while reusing the ASTContext
+          // for the rest of the file. Set the location so that it maps back to
+          // the original file.
+          setLocationInfoForRange(SM, VDRange, VDBufID, Location,
+                                  /*presumed=*/true);
+          // Adjust offset according to generated buffer.
+          auto originalLoc = info->originalSourceRange.getStart();
+          auto originalBufID = SM.findBufferContainingLoc(originalLoc);
+          auto generatedLoc = info->generatedSourceRange.getStart();
+          auto generatedBufID = SM.findBufferContainingLoc(generatedLoc);
+          Location.Offset +=
```

Comment:
I don't see line/column being set at this point, which means it would be the pre-offset-fixup line/col. Which makes me a little concerned that we don't have a sufficient test case for this 😬

Reply:
For replaced function bodies, the line/col is set using `getPresumedLineAndColumnForLoc()` rather than `getLineAndColumnInBuffer()` (following what the original `setLocationInfo()` code did). I'm not exactly sure what the difference is between those but I see some kind of offset handling in `getPresumedLineAndColumnForLoc()`.

Reply:
Here's what the original was doing: https://github.com/swiftlang/swift/blob/00aa22f7b26abeaef60b37fc456d1c5d4f7c336b/tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp#L912-L923

There, the line/col are set with `getLineAndColumnInBuffer()`, and then later overwritten with `getPresumedLineAndColumnForLoc()`.

Reply:
Ah yeah, sorry, ignore me. For some reason I assumed it was using the calculated offset.

> That said, I'm not sure what the difference is between the virtual file's line offset and the offset we compute here.

Hopefully none 🤔

Reply:
😬